src/battle_ai_switch_items.c

```diff
@@ -208,28 +231,81 @@ static bool32 HasBadOdds(u32 battler, bool32 emitResult)
 
             // Switch mon out
             gBattleStruct->AI_monToSwitchIntoId[battler] = PARTY_SIZE;
-            if (emitResult)
-                BtlController_EmitTwoReturnValues(battler, 1, B_ACTION_SWITCH, 0);
-            return TRUE;
+            if (IsSwitchinIdValid(battler, gBattleStruct->AI_monToSwitchIntoId[battler]))
+            {
+                if (emitResult)
+                    BtlController_EmitTwoReturnValues(battler, 1, B_ACTION_SWITCH, 0);
+                return TRUE;
+            }
         }
     }
     return FALSE;
 }
 
-static bool32 ShouldSwitchIfAllBadMoves(u32 battler, bool32 emitResult)
+static bool32 ShouldSwitchIfAllMovesBad(u32 battler, bool32 emitResult)
 {
-    if (AI_DATA->shouldSwitchIfBadMoves & (1u << battler))
+    u32 battlerIndex, moveIndex;
+    bool32 switchOut = FALSE;
+
+    // Consider switching if all moves are worthless to use.
+    if (GetTotalBaseStat(gBattleMons[battler].species) >= 310 // Mon is not weak.
+        && gBattleMons[battler].hp >= gBattleMons[battler].maxHP / 2) // Mon has more than 50% of its HP
     {
-        AI_DATA->shouldSwitchIfBadMoves &= ~(1u << battler);
-        gBattleStruct->AI_monToSwitchIntoId[battler] = AI_DATA->monToSwitchId[battler];
-        if (emitResult)
-            BtlController_EmitTwoReturnValues(battler, BUFFER_B, B_ACTION_SWITCH, 0);
-        return TRUE;
+        s32 cap = AI_THINKING_STRUCT->aiFlags[battler] & (AI_FLAG_CHECK_VIABILITY) ? 95 : 93;
```

Comment:
This could be a more descriptive variable but not blocking a merge because of it.

Reply:
I'll include a rename in the ShouldSwitch refactor PR then :)